DerivationOutputs unable to handle infinity values

https://github.com/cambridge-cares/TheWorldAvatar/blob/98df781780dd408d033b16da878fa7e0a9c9341f/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputs.java#L187

The literal triple added via `addLiteral("http://s", "http://p", Double.POSITIVE_INFINITY)` will be serialised to `{<http://s> <http://p> Infinity.}` when updating derivation outputs in KG, resulting an invalid SPARQL update string. The valid triple should read `{<http://s> <http://p> "Infinity"^^xsd:double.}`. The same behaviour applies to negative infinity values (`Double.NEGATIVE_INFINITY`), i.e. `{<http://s> <http://p> -Infinity.}` should be corrected to `{<http://s> <http://p> "-Infinity"^^xsd:double.}`.


Potential fix:
```java
public void addLiteral(String s, String p, Number o) {
	s = trimIRI(s);
	p = trimIRI(p);
	try {
		checkIfValidIri(Arrays.asList(s, p));
	} catch (Exception e) {
		throw new JPSRuntimeException(INVALID_IRI_FOR_ADDING_LITERAL + Arrays.asList(s, p, o).toString());
	}
	if (o.equals(Double.POSITIVE_INFINITY)) {
		this.addLiteral(s, p, "Infinity", XSD.DOUBLE.toString());
	} else if (o.equals(Double.NEGATIVE_INFINITY)) {
		this.addLiteral(s, p, "-Infinity", XSD.DOUBLE.toString());
	} else {
		this.outputTriples.add(Rdf.iri(s).has(Rdf.iri(p), Rdf.literalOf(o)));
	}
}
```

It turns out that the above fix works for SPARQL update with `INSERT DATA` clause, however, fails in `INSERT {} WHERE {}` clause if updated on blazegaph-backended triple store (**it works for rdf4j-backended tho**, confirmed by @kokfoongcmcl). 

On blazegraph, if one execute below update,
```sparql
INSERT {
  <http://s> <http://p> "INF"^^xsd:double .
}
WHERE {
  ?s ?p ?o.
}
```
it will try to parse the double from the string `"INF"` using `Double.parseDouble(String)`, which will run into the below exception as `"INF"` is not considered a valid value to be a Double in Java (the correct string should be `"Infinity"`):
```
          ERROR: SPARQL-UPDATE: updateStr=INSERT {
    "INF"^^xsd:double .
}
WHERE {
  ?s ?p ?o.
}

java.util.concurrent.ExecutionException: java.lang.NumberFormatException: For input string: "INF"
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at com.bigdata.rdf.sail.webapp.BigdataServlet.submitApiTask(BigdataServlet.java:294)
	at com.bigdata.rdf.sail.webapp.QueryServlet.doSparqlUpdate(QueryServlet.java:460)
	at com.bigdata.rdf.sail.webapp.QueryServlet.doPost(QueryServlet.java:241)
	at com.bigdata.rdf.sail.webapp.RESTServlet.doPost(RESTServlet.java:269)
	at com.bigdata.rdf.sail.webapp.MultiTenancyServlet.doPost(MultiTenancyServlet.java:195)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:652)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:346)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:887)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1684)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NumberFormatException: For input string: "INF"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:543)
	at java.base/java.lang.Double.valueOf(Double.java:506)
	at com.bigdata.rdf.sail.sparql.ASTDeferredIVResolutionInitializer.decode(ASTDeferredIVResolutionInitializer.java:618)
	at com.bigdata.rdf.sail.sparql.ASTDeferredIVResolutionInitializer.getBigdataValue(ASTDeferredIVResolutionInitializer.java:374)
	at com.bigdata.rdf.sail.sparql.ASTDeferredIVResolutionInitializer.process(ASTDeferredIVResolutionInitializer.java:229)
	at com.bigdata.rdf.sail.sparql.Bigdata2ASTSPARQLParser.parseUpdate2(Bigdata2ASTSPARQLParser.java:280)
	at com.bigdata.rdf.sail.webapp.QueryServlet$SparqlUpdateTask.call(QueryServlet.java:533)
	at com.bigdata.rdf.sail.webapp.QueryServlet$SparqlUpdateTask.call(QueryServlet.java:472)
	at com.bigdata.rdf.task.ApiTaskForIndexManager.call(ApiTaskForIndexManager.java:68)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more
```

> https://www.oreilly.com/library/view/xml-schema/0596002521/re63.html

The same exception will also be thrown if `"-INF"` is passed in. Although both `"INF"`  and `"-INF"` are valid values in `xsd:double` (in fact, they seem to be _THE_ valid values for infinity), it seems the easiest way, for now, is to use `"Infinity"` and `"-Infinity"` in stead in below two lines, respectively:

change 
> https://github.com/cambridge-cares/TheWorldAvatar/blob/b3d02f4a830a61119c1dfb87c806d04a40eba91d/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputs.java#L199

to `this.addLiteral(s, p, "Infinity", XSD.DOUBLE.toString());`

and change
> https://github.com/cambridge-cares/TheWorldAvatar/blob/b3d02f4a830a61119c1dfb87c806d04a40eba91d/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputs.java#L201

to `this.addLiteral(s, p, "-Infinity", XSD.DOUBLE.toString()); `

This has been tested work for both blazegraph and rdf4j, the python side is also able to query the added literal and parse it to `inf` using `float('Infinity')`.

What do you think? @sm453

--- a/JPS_BASE_LIB/pom.xml
+++ b/JPS_BASE_LIB/pom.xml
@@ -8,7 +8,7 @@
 
     <!-- Please refer to the Versioning page on TheWorldAvatar wiki for
     details on how version numbers should be selected -->
-    <version>1.20.1</version>
+    <version>1.20.2</version>
 
     <!-- Project Properties -->
     <properties>

--- a/JPS_BASE_LIB/python_wrapper/dev_requirements.txt
+++ b/JPS_BASE_LIB/python_wrapper/dev_requirements.txt
@@ -1,1 +1,2 @@
+testcontainers
 pytest

--- a/JPS_BASE_LIB/python_wrapper/py4jps/__init__.py
+++ b/JPS_BASE_LIB/python_wrapper/py4jps/__init__.py
@@ -1,3 +1,3 @@
 from py4jps.JPSGateway import JPSGateway
 
-__version__ = "1.0.25"
+__version__ = "1.0.26"

--- a/JPS_BASE_LIB/python_wrapper/release_py4jps_to_pypi.sh
+++ b/JPS_BASE_LIB/python_wrapper/release_py4jps_to_pypi.sh
@@ -233,7 +233,7 @@ test_release() {
     else
         $PYTHON_EXEC -m pip install --no-cache-dir --upgrade $PROJECT_NAME
     fi
-    $PYTHON_EXEC -m pip install pytest
+    $PYTHON_EXEC -m pip install pytest testcontainers
 
     run_py4jps_tests $PYTHON_EXEC
 

--- a/JPS_BASE_LIB/python_wrapper/setup.py
+++ b/JPS_BASE_LIB/python_wrapper/setup.py
@@ -2,7 +2,7 @@ from setuptools import setup, find_namespace_packages
 
 setup(
     name='py4jps',
-    version='1.0.25',
+    version='1.0.26',
     author='Daniel Nurkowski',
     author_email='danieln@cmclinnovations.com',
     license='MIT',

--- a/JPS_BASE_LIB/python_wrapper/tests/conftest.py
+++ b/JPS_BASE_LIB/python_wrapper/tests/conftest.py
@@ -1,0 +1,58 @@
+from testcontainers.core.container import DockerContainer
+
+import logging
+import pytest
+import time
+
+logging.getLogger("py4j").setLevel(logging.INFO)
+
+
+# ----------------------------------------------------------------------------------
+# Module-scoped test fixtures
+# ----------------------------------------------------------------------------------
+
+@pytest.fixture(scope="module")
+def initialise_triple_store():
+    # NOTE: requires access to the docker.cmclinnovations.com registry from the machine the test is run on.
+    # For more information regarding the registry, see: https://github.com/cambridge-cares/TheWorldAvatar/wiki/Docker%3A-Image-registry
+    blazegraph = DockerContainer(
+        'docker.cmclinnovations.com/blazegraph_for_tests:1.0.0')
+    # the port is set as 9999 to match with the value set in the docker image
+    blazegraph.with_exposed_ports(9999)
+
+    with blazegraph as container:
+        # Wait some arbitrary time until container is reachable
+        time.sleep(3)
+
+        # Retrieve SPARQL endpoint
+        endpoint = get_endpoint(container)
+
+        yield endpoint
+
+        # Clear logger at the end of the test
+        clear_loggers()
+
+
+# ----------------------------------------------------------------------------------
+# Helper functions
+# ----------------------------------------------------------------------------------
+
+def get_endpoint(docker_container):
+    # Retrieve SPARQL endpoint for temporary testcontainer
+    # endpoint acts as both Query and Update endpoint
+    endpoint = 'http://' + docker_container.get_container_host_ip().replace('localnpipe', 'localhost') + ':' \
+               + docker_container.get_exposed_port(9999)
+    # 'kb' is default namespace in Blazegraph
+    endpoint += '/blazegraph/namespace/kb/sparql'
+    return endpoint
+
+
+# method adopted from https://github.com/pytest-dev/pytest/issues/5502#issuecomment-647157873
+def clear_loggers():
+    """Remove handlers from all loggers"""
+    import logging
+    loggers = [logging.getLogger()] + list(logging.Logger.manager.loggerDict.values())
+    for logger in loggers:
+        handlers = getattr(logger, 'handlers', [])
+        for handler in handlers:
+            logger.removeHandler(handler)

--- a/JPS_BASE_LIB/python_wrapper/tests/test_wrapper.py
+++ b/JPS_BASE_LIB/python_wrapper/tests/test_wrapper.py
@@ -1,3 +1,5 @@
+import random
+import uuid
 import pytest
 from py4jps.resources import JpsBaseLib
 from py4j.java_gateway import GatewayParameters
@@ -89,18 +91,24 @@ def test_fileReading():
     assert "test file1" == file_str
     jpsGW.shutdown()
 
-def test_remoteKGquery():
+def test_tripleStoreQuery(initialise_triple_store):
+    endpoint = initialise_triple_store
     jpsGW = JpsBaseLib()
     jpsGW.launchGateway()
     jpsGW_view = jpsGW.createModuleView()
     jpsGW.importPackages(jpsGW_view,"uk.ac.cam.cares.jps.base.query.*")
 
-    StoreRouter = jpsGW_view.StoreRouter
-    StoreClient = StoreRouter.getStoreClient('http://kb/ontokin', True, False)
-    response = StoreClient.executeQuery(("PREFIX ontokin: <http://www.theworldavatar.com/ontology/ontokin/OntoKin.owl#> \
-                                    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>	SELECT ?mechanismIRI \
-                                    WHERE	{ ?mechanismIRI rdf:type ontokin:ReactionMechanism .} LIMIT 10"))
+    StoreClient = jpsGW_view.RemoteStoreClient(endpoint, endpoint)
+    triples = []
+    for i in range(random.randint(1,100)):
+        triples.append(f"<http://{str(uuid.uuid4())}> <http://{str(uuid.uuid4())}> <http://{str(uuid.uuid4())}>.")
+    triples.append(f"""<http://{str(uuid.uuid4())}> <http://{str(uuid.uuid4())}> "Infinity"^^<http://www.w3.org/2001/XMLSchema#double>.""")
+    triples.append(f"""<http://{str(uuid.uuid4())}> <http://{str(uuid.uuid4())}> "-Infinity"^^<http://www.w3.org/2001/XMLSchema#double>.""")
+    triples.append(f"""<http://{str(uuid.uuid4())}> <http://{str(uuid.uuid4())}> "NaN"^^<http://www.w3.org/2001/XMLSchema#double>.""")
+    StoreClient.executeUpdate(f"INSERT {{ {' '.join(triples)} }} WHERE {{}}")
+    response = StoreClient.executeQuery((f"ASK {{ {' '.join(triples)} }}"))
     response = json.loads(str(response))
+    assert response[0]['ASK']
     jpsGW.shutdown()
 
 def test_javaPythonObjConversion():

--- a/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputs.java
+++ b/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputs.java
@@ -196,9 +196,9 @@ public class DerivationOutputs {
 		if (o instanceof Double) {
 			if (((Double) o).isInfinite()) {
 				if (((Double) o) > 0) {
-					this.addLiteral(s, p, "INF", XSD.DOUBLE.toString());
+					this.addLiteral(s, p, "Infinity", XSD.DOUBLE.toString());
 				} else {
-					this.addLiteral(s, p, "-INF", XSD.DOUBLE.toString());
+					this.addLiteral(s, p, "-Infinity", XSD.DOUBLE.toString());
 				}
 			} else if (((Double) o).isNaN()) {
 				this.addLiteral(s, p, "NaN", XSD.DOUBLE.toString());

--- a/JPS_BASE_LIB/src/test/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputsIntegrationTest.java
+++ b/JPS_BASE_LIB/src/test/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputsIntegrationTest.java
@@ -1,0 +1,153 @@
+package uk.ac.cam.cares.jps.base.derivation;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.eclipse.rdf4j.sparqlbuilder.core.Variable;
+import org.eclipse.rdf4j.sparqlbuilder.core.query.ModifyQuery;
+import org.eclipse.rdf4j.sparqlbuilder.core.query.Queries;
+import org.eclipse.rdf4j.sparqlbuilder.core.query.SelectQuery;
+import org.eclipse.rdf4j.sparqlbuilder.graphpattern.TriplePattern;
+import org.json.JSONArray;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.jupiter.api.AfterAll;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+
+import uk.ac.cam.cares.jps.base.exception.JPSRuntimeException;
+import uk.ac.cam.cares.jps.base.query.RemoteStoreClient;
+
+@Testcontainers
+public class DerivationOutputsIntegrationTest {
+    static RemoteStoreClient storeClient;
+    static String kgUrl;
+    ModifyQuery modify;
+
+    // NOTE: requires access to the docker.cmclinnovations.com registry from the machine the test is run on.
+    // For more information regarding the registry, see: https://github.com/cambridge-cares/TheWorldAvatar/wiki/Docker%3A-Image-registry
+    @Container
+    private static GenericContainer<?> blazegraph = new GenericContainer<>(DockerImageName.parse("docker.cmclinnovations.com/blazegraph_for_tests:1.0.0"))
+            .withExposedPorts(9999); // the port is set as 9999 to match with the value set in the docker image
+
+    @BeforeClass
+    public static void initialise()
+            throws NoSuchMethodException, SecurityException {
+        // create the container in a clean state
+        try {
+            blazegraph.start();
+        } catch (Exception e) {
+            throw new JPSRuntimeException("DerivationOutputsIntegrationTest: Docker container startup failed. Please try running tests again");
+        }
+
+        // initialise all variables to be used
+        kgUrl = "http://" + blazegraph.getHost() + ":" + blazegraph.getFirstMappedPort() + "/blazegraph/namespace/kb/sparql";
+        System.out.println(kgUrl);
+        storeClient = new RemoteStoreClient(kgUrl, kgUrl);
+    }
+
+    @AfterAll
+    public static void stopContainers() {
+        // close containers after all tests
+        if (blazegraph.isRunning()) {
+            blazegraph.stop();
+        }
+    }
+
+    @Test
+    public void testSparqlUpdateNumberLiteral() {
+        ////////////////////////
+        // test SPARQL INSERT //
+        ////////////////////////
+        // create a list of triples to insert
+        List<TriplePattern> outputTriples = generateLiteralTriples();
+
+		// sparql update all new generated outputTriples
+        modify = Queries.MODIFY();
+		outputTriples.stream().forEach(t -> modify.insert(t));
+        System.out.println(modify.getQueryString());
+		storeClient.executeUpdate(modify.getQueryString());
+
+        // check that the triples were inserted
+        Assert.assertTrue(allOutputTriplesExistInKG(outputTriples));
+
+        //////////////////////////////
+        // test SPARQL INSERT-WHERE //
+        //////////////////////////////
+        // create a list of triples to insert
+        outputTriples = generateLiteralTriples();
+
+		// sparql update all new generated outputTriples
+        SelectQuery query = Queries.SELECT();
+        Variable s = query.var();
+        Variable p = query.var();
+        Variable o = query.var();
+        modify = Queries.MODIFY();
+		outputTriples.stream().forEach(t -> modify.insert(t));
+        modify.where(s.has(p, o));
+        System.out.println(modify.getQueryString());
+		storeClient.executeUpdate(modify.getQueryString());
+
+        // check that the triples were inserted
+        Assert.assertTrue(allOutputTriplesExistInKG(outputTriples));
+
+        /////////////////////////////////////
+        // test SPARQL DELETE-INSERT-WHERE //
+        /////////////////////////////////////
+        // create a list of triples to insert
+        outputTriples = generateLiteralTriples();
+
+        // sparql update all new generated outputTriples
+        query = Queries.SELECT();
+        s = query.var();
+        p = query.var();
+        o = query.var();
+        modify = Queries.MODIFY();
+		outputTriples.stream().forEach(t -> modify.insert(t));
+        modify.delete(s.has(p, o)).where(s.has(p, o));
+        System.out.println(modify.getQueryString());
+		storeClient.executeUpdate(modify.getQueryString());
+
+        // check that the triples were inserted
+        Assert.assertTrue(allOutputTriplesExistInKG(outputTriples));
+    }
+
+    public List<TriplePattern> generateLiteralTriples() {
+        DerivationOutputs derivationOutputs = new DerivationOutputs();
+
+        String s1 = "http://" + UUID.randomUUID().toString();
+		String p1 = "http://" + UUID.randomUUID().toString();
+		Double o1 = Double.NaN; // add NaN
+
+		String s2 = "http://" + UUID.randomUUID().toString();
+		String p2 = "http://" + UUID.randomUUID().toString();
+		Double o2 = Double.POSITIVE_INFINITY; // add positive infinity INF
+
+        String s3 = "http://" + UUID.randomUUID().toString();
+		String p3 = "http://" + UUID.randomUUID().toString();
+		Double o3 = Double.NEGATIVE_INFINITY; // add negative infinity -INF
+
+        String s4 = "http://" + UUID.randomUUID().toString();
+		String p4 = "http://" + UUID.randomUUID().toString();
+		Double o4 = 2.3; // add a normal number
+
+        derivationOutputs.addLiteral(s1, p1, o1);
+        derivationOutputs.addLiteral(s2, p2, o2);
+        derivationOutputs.addLiteral(s3, p3, o3);
+        derivationOutputs.addLiteral(s4, p4, o4);
+
+        return derivationOutputs.getOutputTriples();
+    }
+
+    public boolean allOutputTriplesExistInKG(List<TriplePattern> outputTriples) {
+        // check if all outputTriples exist in the triple store
+        String queryString = "ASK { " + outputTriples.stream().map(t -> t.getQueryString()).collect(Collectors.joining("")) + " }";
+        JSONArray queryResult = storeClient.executeQuery(queryString);
+        System.out.println(queryResult);
+        return queryResult.getJSONObject(0).getBoolean("ASK");
+    }
+}

--- a/JPS_BASE_LIB/src/test/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputsTest.java
+++ b/JPS_BASE_LIB/src/test/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputsTest.java
@@ -579,9 +579,9 @@ public class DerivationOutputsTest {
 			if (o instanceof Double) {
 				if (((Double) o).isInfinite()) {
 					if (((Double) o) > 0) {
-						return "<" + s + "> <" + p + "> \"INF\"^^<" + XSD.DOUBLE.toString() + "> .";
+						return "<" + s + "> <" + p + "> \"Infinity\"^^<" + XSD.DOUBLE.toString() + "> .";
 					} else {
-						return "<" + s + "> <" + p + "> \"-INF\"^^<" + XSD.DOUBLE.toString() + "> .";
+						return "<" + s + "> <" + p + "> \"-Infinity\"^^<" + XSD.DOUBLE.toString() + "> .";
 					}
 				} else if (((Double) o).isNaN()) {
 					return "<" + s + "> <" + p + "> \"NaN\"^^<" + XSD.DOUBLE.toString() + "> .";

